src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs

```diff
@@ -1,278 +1,156 @@
-            var tokenAfterAuthenticatedRequest = await tokenCache.GetAsync(invoiceReadClientName, cts.Token);
-            Assert.NotNull(tokenAfterAuthenticatedRequest);
-            Assert.Equal(tokenAfterStartupBackgroundCaching, tokenAfterAuthenticatedRequest);
+        // Creating multiple test APIs helps simulate a distributed system with multiple replicas
+        this._api1 = this.CreateTestApi("app1", this._idp);
+        this._api2 = this.CreateTestApi("app2", this._idp);
 
-            // Wait until token is expired - before that, there should be no background refresh
-            var tokenManagementService = webApp.Services.GetRequiredService<ClientCredentialsTokenManagementService>();
-            Assert.Equal(0, tokenManagementService.BackgroundRefreshedTokenCount);
-            await Task.Delay(tokenAfterAuthenticatedRequest.Expiration - DateTimeOffset.UtcNow, cts.Token);
+        // Ensure that registered clients get cached tokens on app startup
+        await this._api1.StartAsync(this._timeout.Token);
+        await this._api1.Services.GetRequiredService<OnStartupTokenCacheBackgroundService>().WaitForTokenCachingToCompleteAsync(this._timeout.Token);
 
-            // At this point we're not making any new authenticated HTTP request,
-            // but a background task should have refreshed the token, which should differ from the initially cached one
-            Assert.Equal(1, tokenManagementService.BackgroundRefreshedTokenCount);
-            var tokenAfterFirstBackgroundRefresh = await tokenCache.GetAsync(invoiceReadClientName, cts.Token);
-            Assert.NotNull(tokenAfterFirstBackgroundRefresh);
-            Assert.NotEqual(tokenAfterStartupBackgroundCaching, tokenAfterFirstBackgroundRefresh);
+        await this._api2.StartAsync(this._timeout.Token);
+        await this._api2.Services.GetRequiredService<OnStartupTokenCacheBackgroundService>().WaitForTokenCachingToCompleteAsync(this._timeout.Token);
 
-            // If we wait a little bit longer, the token should be refreshed again
-            await Task.Delay(tokenAfterFirstBackgroundRefresh.Expiration - DateTimeOffset.UtcNow, cts.Token);
-            Assert.Equal(2, tokenManagementService.BackgroundRefreshedTokenCount);
-            var tokenAfterSecondBackgroundRefresh = await tokenCache.GetAsync(invoiceReadClientName, cts.Token);
-            Assert.NotNull(tokenAfterSecondBackgroundRefresh);
-            Assert.NotEqual(tokenAfterFirstBackgroundRefresh, tokenAfterSecondBackgroundRefresh);
-        }
-        finally
-        {
-            // Shut down the web app
-#pragma warning disable CA1849 // Given that we are using .Net6/7/8, updating the method to CancelAsync would break backwards compatibility
-            cts.Cancel();
-#pragma warning restore CA1849
-        }
+        // TODO improve background token acquisition so that 2nd app uses the token cached by the 1st app from the distributed cache
+        Assert.Equal(2, this._tokenRequestCount);
```

Comment:
This is where we have to work on in the upcoming improvement story. When two apps start one after another and both try to pre-fetch the token, the IDP server should have received a single token request, not two.

It means the token refresh process is too agressive and we don't leverage the L2 cache.

Reply:
Is it a possible race condition: sometimes if one of the process is a bit slower it will have the time to be in the L2 cache?